test/unit/contrib/parallel/buffer_queue_parallel_test.cpp

```diff
@@ -189,9 +190,10 @@ void test_buffer_queue_wait_throw(size_t initialCapacity)
                         push_status = ex;
                     }
                 }
-               // printf("stop writer #%ld %lu\n", tid, offset_end - offset);
+                // printf("stop writer #%ld %lu\n", tid, offset_end - offset);
                 // Last writer! No more values will come, so we close the queue.
-                if (registered_writer.fetch_sub(1) == 1)
+                ++registered_writer;
```

Comment:
Yes, this is necessary and the actual root of the dead lock. Otherwise I had to introduce a separate counting variable. The reason is, that the reader threads wait for the writers to register and the other way around. In order to guarantee at least some overlap between concurrent write and read. So the reader wait for all writers to register. The same do the writers. But, there might be the risk that between the registered_reader being incremented and checking if all writers are registered one writer might have finished it's job and already decremented the writer count. Thus the reader will wait forever. By inverting the decrement to an increment here we avoid this dead lock.